types/wlr_relative_pointer_v1.c

```diff
@@ -0,0 +1,234 @@
+		relative_pointer);
+
+	wlr_log(WLR_DEBUG, "relative_pointer_v1 %p created for client %p",
+		relative_pointer, client);
+}
+
+
+static void relative_pointer_manager_v1_bind(struct wl_client *wl_client, void *data,
+		uint32_t version, uint32_t id) {
+	struct wlr_relative_pointer_manager_v1 *relative_pointer_manager = data;
+
+	struct wl_resource *manager_resource = wl_resource_create(wl_client,
+		&zwp_relative_pointer_manager_v1_interface, version, id);
+
+	if (manager_resource == NULL) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	wl_list_insert(&relative_pointer_manager->resources, wl_resource_get_link(manager_resource));
+
+	wl_resource_set_implementation(manager_resource, &relative_pointer_manager_v1_impl,
+		relative_pointer_manager, relative_pointer_manager_v1_handle_resource_destroy);
+
+	wlr_log(WLR_DEBUG, "relative_pointer_v1 manager bound to client %p",
+		wl_client);
+}
+
+static void handle_display_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_relative_pointer_manager_v1 *relative_pointer_manager =
+		wl_container_of(listener, relative_pointer_manager, display_destroy_listener);
+	wlr_relative_pointer_manager_v1_destroy(relative_pointer_manager);
+}
+
+
+/**
+ * Implementations
+ */
+
+static const struct zwp_relative_pointer_manager_v1_interface relative_pointer_manager_v1_impl = {
+	.destroy = relative_pointer_manager_v1_handle_destroy,
+	.get_relative_pointer = relative_pointer_manager_v1_handle_get_relative_pointer,
+};
+
+
+static const struct zwp_relative_pointer_v1_interface relative_pointer_v1_impl = {
+	.destroy = relative_pointer_v1_handle_destroy,
+};
+
+
+/**
+ * Public functions
+ */
+
+struct wlr_relative_pointer_manager_v1 *wlr_relative_pointer_manager_v1_create(struct wl_display *display) {
+	struct wlr_relative_pointer_manager_v1 *relative_pointer_manager =
+		calloc(1, sizeof(struct wlr_relative_pointer_manager_v1));
+
+	if (relative_pointer_manager == NULL) {
+		return NULL;
+	}
+
+	wl_list_init(&relative_pointer_manager->resources);
+	wl_list_init(&relative_pointer_manager->relative_pointers);
+
+	relative_pointer_manager->global = wl_global_create(display,
+		&zwp_relative_pointer_manager_v1_interface, RELATIVE_POINTER_MANAGER_VERSION,
+		relative_pointer_manager, relative_pointer_manager_v1_bind);
+
+	if (relative_pointer_manager->global == NULL) {
+		free(relative_pointer_manager);
+		return NULL;
+	}
+
+	wl_signal_init(&relative_pointer_manager->events.destroy);
+	wl_signal_init(&relative_pointer_manager->events.new_relative_pointer);
+
+	relative_pointer_manager->display_destroy_listener.notify = handle_display_destroy;
+	wl_display_add_destroy_listener(display, &relative_pointer_manager->display_destroy_listener);
+
+	wlr_log(WLR_DEBUG, "relative_pointer_v1 manager created");
+
+	return relative_pointer_manager;
+}
+
+
+void wlr_relative_pointer_manager_v1_destroy(struct wlr_relative_pointer_manager_v1 *relative_pointer_manager) {
```

Comment:
Style: `relative_pointer_manager` could just be named `manager`, it's clear what we're talking about since we're in `wlr_relative_pointer_v1.c`.

Reply:
Bump

Reply:
Sorry about that.  I thought I did this with my last commit.